packages/frontend-test-utils/src/app/createExtensionTester.tsx

```diff
@@ -0,0 +1,300 @@
+  },
+  output: {
+    element: coreExtensionData.reactElement,
+  },
+  factory({ inputs }) {
+    const NavItem = (props: {
+      routeRef: RouteRef<undefined>;
+      title: string;
+      icon: IconComponent;
+    }) => {
+      const { routeRef, title, icon: Icon } = props;
+      const to = useRouteRef(routeRef)();
+      return (
+        <li>
+          <Link to={to}>
+            <Icon /> {title}
+          </Link>
+        </li>
+      );
+    };
+    return {
+      element: (
+        <nav>
+          <ul>
+            {inputs.items.map((item, index) => (
+              <NavItem
+                key={index}
+                icon={item.output.target.icon}
+                title={item.output.target.title}
+                routeRef={item.output.target.routeRef}
+              />
+            ))}
+          </ul>
+        </nav>
+      ),
+    };
+  },
+});
+
+const AuthenticationProvider = (props: {
+  signInPage?: ComponentType<SignInPageProps>;
+  children: ReactNode;
+}) => {
+  const { signInPage: SignInPage, children } = props;
+  const configApi = useApi(configApiRef);
+  const signOutTargetUrl = getBasePath(configApi) || '/';
+
+  const internalAppContext = useContext(InternalAppContext);
+  if (!internalAppContext) {
+    throw new Error('AppRouter must be rendered within the AppProvider');
+  }
+
+  const { appIdentityProxy } = internalAppContext;
+  const [identityApi, setIdentityApi] = useState<IdentityApi>();
+
+  if (!SignInPage) {
+    appIdentityProxy.setTarget(
+      {
+        getUserId: () => 'guest',
+        getIdToken: async () => undefined,
+        getProfile: () => ({
+          email: 'guest@example.com',
+          displayName: 'Guest',
+        }),
+        getProfileInfo: async () => ({
+          email: 'guest@example.com',
+          displayName: 'Guest',
+        }),
+        getBackstageIdentity: async () => ({
+          type: 'user',
+          userEntityRef: 'user:default/guest',
+          ownershipEntityRefs: ['user:default/guest'],
+        }),
+        getCredentials: async () => ({}),
+        signOut: async () => {},
+      },
+      { signOutTargetUrl },
+    );
+
+    return children;
+  }
+
+  if (!identityApi) {
+    return <SignInPage onSignInSuccess={setIdentityApi} />;
+  }
+
+  appIdentityProxy.setTarget(identityApi, {
+    signOutTargetUrl,
+  });
+
+  return children;
+};
+
+const TestCoreRouterExtension = createExtension({
+  namespace: 'core',
+  name: 'router',
+  attachTo: { id: 'core', input: 'root' },
+  inputs: {
+    signInPage: createExtensionInput(
+      {
+        component: createSignInPageExtension.componentDataRef,
+      },
+      { singleton: true, optional: true },
+    ),
+    children: createExtensionInput(
+      {
+        element: coreExtensionData.reactElement,
+      },
+      { singleton: true },
+    ),
+  },
+  output: {
+    element: coreExtensionData.reactElement,
+  },
+  factory({ inputs }) {
+    const SignInPage = inputs.signInPage?.output.component;
```

Comment:
> [!Note]
> The sign-in page is used for fixing [this](https://github.com/backstage/backstage/actions/runs/7200236374/job/19613774294?pr=21782) test.

Reply:
Alright, in the end it's probably worth it rather than having a separate harness for testing sign-in pages.